### PR TITLE
LA-737 Sign up for SaaS

### PR DESCRIPTION
**Pull request by dab246**
https://user-images.githubusercontent.com/80730648/143542791-a0796620-df43-4c13-a1ad-8d8ed6d8c78a.mov

### CODE REVIEW

File: domain/lib/src/model/authentication/secret_token.dart [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+// <http://www.gnu.org/licenses/> for the GNU Affero General Public License version
+//  3 and <http://www.linshare.org/licenses/LinShare-License_AfferoGPL-v3.pdf> for
+//  the Additional Terms applicable to LinShare software.
+
+import 'package:equatable/equatable.dart';
+
+class SecretToken with EquatableMixin {
+  final String value;
+
+  SecretToken(this.value);
```
Comment (hoangdat) [inline domain/lib/src/model/authentication/secret_token.dart:37]: `SaaSSecretToken`

File: data/lib/src/network/linshare_http_client.dart [hunk 2]
```diff
@@ -734,11 +737,24 @@ class LinShareHttpClient {
...
 
   Future<String> getOIDCConfiguration(Uri baseUrl) async {
-    final result = await _dioClient.get(
-        Endpoint.oidcConfiguration.generateOIDCConfigurationUrl(baseUrl));
+    final result = await _dioClient.get(Endpoint.oidcConfiguration.generateBaseUrl(baseUrl));
     return result;
   }
 
+  Future<SecretTokenResponse> getSecretToken(Uri baseUrl, PlanBodyRequest planBodyRequest) async {
+    final result = await _dioClient.post(
```
Comment (hoangdat) [inline data/lib/src/network/linshare_http_client.dart:745]: we can use the same _dioClient for getting `secretToken` from other url

Comment (dab246) [reply data/lib/src/network/linshare_http_client.dart:745]: Of course !

**Comment by hoangdat** [comment]
```
class PlanRequest with EquatableMixin {
  final String planId;

  PlanRequest(this.planId);

  factory PlanRequest.free() => PlanRequest('free');

  @override
  List<Object> get props => [planId];
}
```

**Comment by dab246** [comment]
@hoangdat  please review

File: data/lib/src/network/linshare_http_client.dart [hunk 3]
```diff
@@ -734,11 +737,32 @@ class LinShareHttpClient {
...
+    return VerifyEmailResponse.fromJson(result);
+  }
+
+  Future<UserSaaSResponse> signUpForSaaS(Uri baseUrl, SignUpBodyRequest signUpBodyRequest) async {
+    final result = await _dioClient.post(
+        Endpoint.signUp.generateBaseUrl(baseUrl),
+        data: signUpBodyRequest.toJson());
+    return UserSaaSResponse.fromJson(result);
+  }
+
```
Comment (hoangdat) [inline data/lib/src/network/linshare_http_client.dart:765]: How about separate three of it to SaaSHttpClient?

**Review by hoangdat — APPROVED** [review]

File: data/lib/src/network/linshare_http_client.dart (hunk 3 above)
Comment (dab246) [reply data/lib/src/network/linshare_http_client.dart:769]: Please see [https://github.com/linagora/linshare-mobile-flutter-app/pull/759/commits/5caa903973270a1cf29dbecfb577794314d5d7ae ](https://github.com/linagora/linshare-mobile-flutter-app/pull/759/commits/5caa903973270a1cf29dbecfb577794314d5d7ae )

File: data/lib/src/repository/authentication/saas_console_repository_impl.dart [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+// <http://www.gnu.org/licenses/> for the GNU Affero General Public License version
+//  3 and <http://www.linshare.org/licenses/LinShare-License_AfferoGPL-v3.pdf> for
+//  the Additional Terms applicable to LinShare software.
+
+import 'package:domain/domain.dart';
+
+class SaaSConsoleRepositoryImpl extends SaaSConsoleRepository {
+
+  final loginBaseUrl = 'https://user.linshare-saas-on-commit.integration-linshare.org';
+  final secretBaseUrl = 'http://subscription-twakeconsole.dev-linshare.lin-saas.dev';
```
Comment (hoangdat) [inline data/lib/src/repository/authentication/saas_console_repository_impl.dart:37]: `https`

File: data/lib/src/datasource_impl/authentication_saas_datasource.dart [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+  AuthenticationSaaSDataSource(this.saaSHttpClient, this.remoteExceptionThrower);
+
+  @override
+  Future<Token> createPermanentTokenWithOIDC(Uri baseUrl, TokenOIDC tokenOIDC, {OTPCode? otpCode}) {
+    throw UnimplementedError();
+  }
+
+  @override
+  Future<OIDCConfiguration?> getOIDCConfiguration(Uri baseUrl) {
```
Comment (hoangdat) [inline data/lib/src/datasource_impl/authentication_saas_datasource.dart:49]: why SaaS don't need to get OIDC Configuration?

Comment (dab246) [reply data/lib/src/datasource_impl/authentication_saas_datasource.dart:49]: Yes. it is shared with oidc.

**Review by hoangdat — APPROVED** [review]

**Pull request merged**